ui/ui-components/pages/_app.tsx

```diff
@@ -10,27 +11,51 @@ import PageTransition from "../components/PageTransition";
 import StatusSubscription from "../components/StatusSubscription";
 import "../components/Icons";
 
+import { PageContext } from "../contexts/page";
+
 import { Actions } from "../utils/apiData";
-import * as eventHandlers from "../utils/eventHandlers";
+import * as eventHandlers from "../eventHandlers";
 import { getModelFromUrlOrCookie, setModelCookie } from "../utils/modelHelper";
 
-export default function GrouparooWebApp(props) {
+export interface GrouparooNextAppProps {
+  clusterName: Actions.NavigationList["clusterName"];
+  currentTeamMember: Partial<Actions.NavigationList["teamMember"]>;
+  hydrationError?: string;
+  navigation: Actions.NavigationList["navigation"];
+  navigationMode: Actions.NavigationList["navigationMode"];
+  navigationModel: Actions.NavigationList["navigationModel"];
+}
+
+export default function GrouparooNextApp(
+  props: AppProps & GrouparooNextAppProps & { err: any }
+) {
   const { Component, pageProps, err, hydrationError } = props;
 
-  const combinedProps = Object.assign({}, pageProps || {}, {
+  const combinedProps = {
+    ...pageProps,
+    ...eventHandlers,
     navigation: props.navigation,
     navigationMode: props.navigationMode,
     navigationModel: props.navigationModel,
     clusterName: props.clusterName,
     currentTeamMember: props.currentTeamMember,
-    ...eventHandlers,
-  });
+  };
+
+  const pageContext = useMemo<PageContext>(() => {
+    return {
+      navigation: props.navigation,
+      navigationMode: props.navigationMode,
+      navigationModel: props.navigationModel,
+      clusterName: props.clusterName,
+      currentTeamMember: props.currentTeamMember,
+    };
+  }, [props]);
 
   return (
-    <Injection {...combinedProps}>
+    <Injection contextValues={[[PageContext, pageContext]]}>
```

Comment:
I think I see what you're doing here, though I'm partial to [this](https://github.com/grouparoo/cloud/blob/main/web/contexts/ContextProvider.component.tsx)

Reply:
Yep, it's pretty much doing the same thing. I could eliminate injection entirely. I don't think it's necessary to have it and have my helper function provide all the contexts. The reason I use the function is to avoid having another nested component in the React inspector.